mdetsims/metacal/metacal_fitter.py

```diff
@@ -0,0 +1,284 @@
+    def _do_all_metacal(self, mbobs_list, data=None):
+        """run metacal on all objects
+
+        NOTE: failed mbobs will have no entries in the final list
+        """
+
+        assert len(mbobs_list[0]) == self.nband
+
+        datalist = []
+        for i, mbobs in enumerate(mbobs_list):
+            if self._check_flags(mbobs):
+                try:
+                    boot = self._do_one_metacal(mbobs)
+                    if isinstance(boot, dict):
+                        res = boot
+                    else:
+                        res = boot.get_metacal_result()
+                except (BootPSFFailure, BootGalFailure) as err:
+                    logger.debug(str(err))
+                    res = {'mcal_flags': 1}
+
+                if res['mcal_flags'] != 0:
+                    logger.debug("metacal fit failed")
+                else:
+                    # make sure we send an array
+                    fit_data = self._get_metacal_output(res, self.nband, mbobs)
+                    if data is not None:
+                        odata = data[i:i+1]
+                        fit_data = eu.numpy_util.add_fields(
+                            fit_data,
+                            odata.dtype.descr,
+                        )
+                        eu.numpy_util.copy_fields(odata, fit_data)
+
+                    self._print_result(fit_data)
+                    datalist.append(fit_data)
+
+        if len(datalist) == 0:
+            return None
+
+        output = eu.numpy_util.combine_arrlist(datalist)
+        return output
+
+    def _do_one_metacal(self, mbobs):
+        conf = self['metacal']
+        psf_pars = conf['psf']
+        max_conf = conf['max_pars']
+
+        tpsf_obs = mbobs[0][0].psf
+        if not tpsf_obs.has_gmix():
+            _fit_one_psf(tpsf_obs, psf_pars)
+
+        psf_Tguess = tpsf_obs.gmix.get_T()
+
+        boot = self._get_bootstrapper(mbobs)
+        if 'lm_pars' in psf_pars:
+            psf_fit_pars = psf_pars['lm_pars']
+        else:
+            psf_fit_pars = None
+
+        prior = self.metacal_prior
+        guesser = None
+
+        boot.fit_metacal(
+            psf_pars['model'],
+            conf['model'],
+            max_conf['pars'],
+            psf_Tguess,
+            psf_fit_pars=psf_fit_pars,
+            psf_ntry=psf_pars['ntry'],
+            prior=prior,
+            guesser=guesser,
+            ntry=max_conf['ntry'],
+            metacal_pars=conf['metacal_pars'],
+        )
+        return boot
+
+    def _check_flags(self, mbobs):
+        """
+        only one epoch, so anything that hits an edge
+        """
+        flags = self['metacal'].get('bmask_flags', None)
+
+        if flags is not None:
+            for obslist in mbobs:
+                for obs in obslist:
+                    msk = (obs.bmask & flags) != 0
+                    if np.any(msk):
+                        logger.info("   EDGE HIT")
+                        return False
+
+        return True
```

Comment:
I slightly redid the logic here to return immediately if anything comes up false.

Reply:
for multi-epoch that's not good. Its ok for one of the epochs to be tossed

Reply:
Hmmmmm. Well the old code did the same thing as the code here so I am guessing it was only for single epoch too.

Reply:
that's true, I was not testing multi-epoch